% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ReadVCF.R
\name{ReadVCF}
\alias{ReadVCF}
\title{Read  vcf file}
\usage{
ReadVCF(filename = NULL, availmemGb = 16, quiet = TRUE)
}
\arguments{
\item{filename}{contains the name of the vcf  file. The file name needs to be in quotes. If the file is not in the working directory, then the full path 
to the file is required.}

\item{availmemGb}{a numeric value. It specifies the amount of available memory (in Gigabytes). 
This should be set to be as large as possible for best performance.}

\item{quiet}{a logical value. If set to \code{TRUE}, additional runtime output is printed.}
}
\value{
To allow Eagle to handle data larger than the memory capacity of a machine, \code{ReadVCF} doesn't load 
the marker data into memory. Instead, it 
writes a reformatted version of the marker data, and its transpose, to the harddrive. These two files
are only temporary, being removed at the end of the R session. 
The object returned by
\code{ReadVCF} is a list object with the elements \code{tmpM} , \code{tmpMt}, and \code{dim_of_M}  
which is the full file name (name and path)  of the reformatted file for the marker  data,  the full file name of the reformatted file 
for the transpose of the marker  data,  and a 2 element vector with the first element the number of individuals and the second 
element the number of marker loci.
}
\description{
A function for reading in marker data in vcf format.
}
\details{
VCF is a tab separated text file containing  meta-information lines, a header line, and then data 
lines. The data lines  contain information about a position in the genome.

It is assumed that genotype information has been recorded on samples for each position. 

Loci with more than two alleles will be removed automatically. 

Eagle will only accept a single (uncompressed) vcf file. If chromosomal information has been recorded in separate 
vcf files, these files need to be merged into a single vcf file.  This can be done by using the 
 BCFtools utility set with command line "bcftools concat".
}
\examples{
  #
  # Read in the genotype data contained in the vcf file geno.vcf
  #
  # The function system.file() gives the full file name (name + full path).
  complete.name <- system.file('extdata', 'geno.vcf', package='Eagle')
  # 
  # The full path and name of the file is
  print(complete.name)
  
  # The file contains 5 marker loci recorded on 3 individuals
  # Two of the loci contain multiple alleles and are removed. 
  # A summary of the file is printed once the file has been read.
  geno_obj <- ReadVCF(filename=complete.name, availmemGb=4) 
   
  # view list contents of geno_obj
  print(geno_obj)


}
